pkg/didcomm/crypto/jwe/authcrypt/encrypt.go

```diff
@@ -0,0 +1,368 @@
+// Using Chacha20Poly1035 encryption/authentication
+
+// C20P Chacha20Poly1035 algorithm
+const C20P = "C20P"
+
+// XC20P XChacha20Poly1035 algorithm
+const XC20P = "XC20P"
+
+// RandReader is a cryptographically secure random number generator.
+var RandReader = rand.Reader
+
+type keyPair struct {
+	priv *[chacha20poly1305.KeySize]byte
+	pub  *[chacha20poly1305.KeySize]byte
+}
+
+// Crypter represents an Authcrypt Encrypter (Decrypter) that outputs/reads JWE envelopes
+type Crypter struct {
+	sender     keyPair
+	recipients []*[chacha20poly1305.KeySize]byte
+	alg        string
+	nonceSize  int
+}
+
+// Envelope represents a JWE envelope as per the Aries Encryption envelop specs
+type Envelope struct {
+	Protected  string      `json:"protected,omitempty"`
+	Recipients []Recipient `json:"recipients,omitempty"`
+	Aad        string      `json:"aad,omitempty"`
+	Iv         string      `json:"iv,omitempty"`
+	Tag        string      `json:"tag,omitempty"`
+	CipherText string      `json:"ciphertext,omitempty"`
+}
+
+// jweHeaders are the Protected JWE headers in a map format
+type jweHeaders map[string]string
+
+// Recipient is a recipient of an envelop including the shared encryption key
+type Recipient struct {
+	EncryptedKey string           `json:"encrypted_key,omitempty"`
+	Header       RecipientHeaders `json:"header,omitempty"`
+}
+
+// RecipientHeaders are the recipient headers
+type RecipientHeaders struct {
+	Apu string `json:"apu,omitempty"`
+	Iv  string `json:"iv,omitempty"`
+	Tag string `json:"tag,omitempty"`
+	Kid string `json:"kid,omitempty"`
+	Oid string `json:"oid,omitempty"`
+}
+
+// New will create an encrypter instance to 'AuthCrypt' payloads for the given sender and recipients arguments
+// and the encryption alg argument. Possible algorithms supported are:
+// C20P (chacha20-poly1035 ietf)
+// XC20P (xchacha20-poly1035 ietf)
+// The returned crypter contains all the information required to encrypt payloads.
+func New(sender keyPair, recipients []*[chacha20poly1305.KeySize]byte, alg string) (*Crypter, error) {
+	var nonceSize int
+	switch alg {
+	case C20P:
+		nonceSize = chacha20poly1305.NonceSize
+	case XC20P:
+		nonceSize = chacha20poly1305.NonceSizeX
+	default:
+		return nil, errors.New(fmt.Sprintf("encryption algorithm '%s' not supported", alg))
+	}
+	if len(recipients) == 0 {
+		return nil, errors.New("empty recipients keys, must have at least one recipient")
+	}
+	var recipientsKey []*[chacha20poly1305.KeySize]byte
+	recipientsKey = append(recipientsKey, recipients...)
+
+	c := &Crypter{
+		sender,
+		recipientsKey,
+		alg,
+		nonceSize,
+	}
+
+	if !isKeyPairValid(sender) {
+		return nil, errors.New(fmt.Sprintf("sender keyPair not supported, it must have %d bytes keys", chacha20poly1305.KeySize))
+	}
+
+	return c, nil
+}
+
+func isKeyPairValid(kp keyPair) bool {
+	if kp.priv == nil || kp.pub == nil || len(kp.priv) != chacha20poly1305.KeySize || len(kp.pub) != chacha20poly1305.KeySize {
```

Comment:
It doesn't seem like this needs to be such a long line.
https://github.com/golang/go/wiki/CodeReviewComments#line-length

Reply:
reduced... will remove the len() check as it's not possible (I thought the linter will catch it)

Reply:
I had meant break it into multiple lines, but dead code removal is even better ;).

Reply:
done :)